Adding Axiom for Immutable References

paaassscccaaalll opened the pull request:
Adds this axiom:

```
axiom s_Ref_immutable_ax_ref_immutable_cons_typeof {
    (forall r: Ref, p: s_Param ::
      { s_Ref_immutable_typeof(s_Ref_immutable_cons(r, p)) }
      s_Ref_immutable_typeof(s_Ref_immutable_cons(r, p)) ==
      s_Ref_immutable_type(typ(p)))
  }
```
to `domain s_Ref_immutable`.

JonasAlaif commented on prusti-encoder/src/encoders/type/kinds/immref.rs:
```diff
@@ -24,17 +24,44 @@ pub(crate) fn domain<'vir>(
+        vir::ViperIdent::new("s_Ref_immutable_type"),
+        vir::UnknownArity::new(builder.vcx.alloc_slice(&[type_type])),
+        type_type,
+    );
```
These all exist in the context, please do not recreate them

JonasAlaif commented:
Also please make sure that the branch you are merging in has been rebased on top of the main `Aurel300:rewrite-2023` one: there should only be a single commit that shows up in this PR

paaassscccaaalll committed:
new axiom for immref

added my functionality to up to date branch

deleted pcs

deleted utils and put helper function into

change is_ad_trusted() to is_type_trusted() which now take in a ty:TY

not using a filter anymore

formating fix PR 68

Update prusti-encoder/src/encoders/mod.rs

removed comment

Co-authored-by: Aurel <Aurel300@users.noreply.github.com>

removed Trusted from Domain

moved trusted handeling to predicate.rs

changed prusti_rustc_interface::middle::ty::Ty to use ty::Ty directly

style improvement

paaassscccaaalll committed:
information from context

Aurel300 commented on prusti-encoder/src/encoders/type/kinds/immref.rs:
```diff
@@ -29,12 +31,19 @@ pub(crate) fn domain<'vir>(
     builder.axiom("value", vir::expr! {
         forall r: [prim_type], value: [inner_type] :: {[cons_ident](r, value)} ([value_ident]([cons_ident](r, value))) == (value)
     });
+    builder.axiom("ref_immutable_cons_typeof", vir::expr! {
```
```suggestion
    builder.axiom("typeof", vir::expr! {
```
(the names of axioms built using the builder are all prefixed with the type name already)

Aurel300 committed:
Update prusti-encoder/src/encoders/type/kinds/immref.rs
